reconcile/utils/oc.py

```diff
@@ -668,6 +682,252 @@ def _run_json(self, cmd, allow_not_found=False):
+                [p['metadata']['name']
+                 for p
+                 in self.get_all('Project.project.openshift.io')['items']]
+        self.init_api_resources = init_api_resources
+        if self.init_api_resources:
+            self.api_resources = self.api_kind_version.keys()
+        else:
+            self.api_resources = None
+
+    def _get_client(self, server, token):
+        opts = dict(
+            api_key={'authorization': f'Bearer {token}'},
+            host=server,
+            verify_ssl=False,
+            # default timeout seems to be 1+ minutes and it can't
+            # be changed, so do no retries on connection failure
+            retries=0
+        )
+
+        if self.jump_host is not None:
+            # the ports could be parameterized, but at this point
+            # we only have need of 1 tunnel for 1 service
+            self.jump_host.create_ssh_tunnel(8888, 8888)
+            opts['proxy'] = 'http://localhost:8888'
+
+        configuration = kubernetes.client.Configuration()
+
+        # the kubernetes client configuration takes a limited set
+        # of parameters during initialization, but there are a lot
+        # more options that can be set to tweak the behavior of the
+        # client via instance variables.  We define a set of options
+        # above in the format of var_name:value then set them here
+        # in the configuration object with setattr.
+        for k, v in opts.items():
+            setattr(configuration, k, v)
+
+        k8s_client = kubernetes.client.ApiClient(configuration)
+        try:
+            return DynamicClient(k8s_client)
+        except urllib3.exceptions.MaxRetryError as e:
+            raise StatusCodeError(f"[{self.server}]: {e}")
+
+    def _get_obj_client(self, kind, group_version):
+        key = f'{kind}.{group_version}'
+        if key not in self.object_clients:
+            self.object_clients[key] = self.client.resources.get(
+                api_version=group_version, kind=kind)
+        return self.object_clients[key]
+
+    def _parse_kind(self, kind_name):
+        kind_group = kind_name.split('.', 1)
+        kind = kind_group[0]
+        if kind in self.api_kind_version:
+            group_version = self.api_kind_version[kind][0]
+        else:
+            raise StatusCodeError(f'{self.server}: {kind} does not exist')
+
+        # if a kind_group has more than 1 entry than the kind_name is in
+        # the format kind.apigroup.  Find the apigroup/version that matches
+        # the apigroup passed with the kind_name
+        if len(kind_group) > 1:
+            apigroup_override = kind_group[1]
+            for gv in self.api_kind_version[kind]:
+                if apigroup_override in gv:
+                    group_version = gv
+                    break
+        return(kind, group_version)
+
+    # this function returns a kind:apigroup/version map for each kind on the
+    # cluster
+    def get_api_resources(self):
```

Comment:
We agreed that the native client would not require tests because we agreed on that once this code was already in place, but the logic of this function is particularly hard to follow. Could I convince you to do some quick unit tests for this method? I believe you would need to mock `client.resources.parse_api_groups` and `client_resources.get_resources_for_api_version` with several values and ensure that ensure that the kind_groupversion return is the one expected.

Reply:
Done